### PR TITLE
feat: RPC - `getTransactionByHash` - `getTransactionReceipt`

### PR DESCRIPTION
**Pull request by glihm**
# Pull Request type

Please check the type of change your PR introduces:

- [ ] Bugfix
- [X] Feature
- [ ] Code style update (formatting, renaming)
- [ ] Refactoring (no functional changes, no API changes)
- [ ] Build-related changes
- [ ] Documentation content changes
- [ ] Testing
- [ ] Other (please describe):

# What is the current behavior?

Issue Number: #192 

# What is the new behavior?

Add support for `starknet_getTransactionByHash`.
This work is on progress for several reasons:
- Transaction receip is expected to be returned, but not yet present in the block.
- Still working on how data are gathered from the runtime to be sent to substrate client.

# Does this introduce a breaking change?

- [ ] Yes
- [X] No

New RPC endpoint.

# Other information

I will add some comment on the code, and I'll keep you posted for new content on the PR or if I am stuck.

### CODE REVIEW

File: crates/client/rpc-core/src/lib.rs
Comment (glihm) [inline crates/client/rpc-core/src/lib.rs:26]: Could it be better to propose a `pub type Hash = FieldElement` to not have `TransactionHash`, `BlockHash`, etc...?

File: crates/client/rpc-core/src/lib.rs
Comment (glihm) [inline crates/client/rpc-core/src/lib.rs:261]: I did a new struct to encapsulate internal data and be sure we can have loose coupling between the struct sent in the RPC response, and internal organization. Is that ok?

File: crates/client/rpc/src/errors.rs
Comment (glihm) [inline crates/client/rpc/src/errors.rs:43]: I propose a new `InvalidHash` that may be more generic for all future Hash related arguments. What do you think?

File: crates/client/rpc/src/lib.rs
Comment (glihm) [inline crates/client/rpc/src/lib.rs:229]: Can you confirm that due to the subtrate architecture, this lines of codes are required to do a runtime call? Because several endpoints will not have the block as argument, and we can then make a thin wrapper to abstract this to them.

File: crates/runtime/src/lib.rs
Comment (glihm) [inline crates/runtime/src/lib.rs:298]: Can you confirm that the block only contains `Transactions` and not the receips. Am I able to find receips directly going into the storage, or for now, I can stick with transactions and I'll update this once receips are present in the block?

File: tests/tests/test-rpc/test-starknet-rpc.ts
Comment (glihm) [inline tests/tests/test-rpc/test-starknet-rpc.ts:86]: I have an error here:

```
LibraryError: -32602: missing field `tx_hash` at line 1 column 27
```

I am not sure what I am missing into the `.ts` file.

**Review by glihm — COMMENTED** [review]
Thank you for your time and answers. :pray:

File: crates/runtime/src/lib.rs
Comment (glihm) [reply crates/runtime/src/lib.rs:298]: Sorry, I just seen the other PR for transaction receipt. I'll focus on Transaction.

File: crates/client/rpc/src/lib.rs
Comment (glihm) [inline crates/client/rpc/src/lib.rs:583]: The mapping here is really heavy. I am looking for alternative, if any. And add the TxType rework if I can as suggested by @EvolveArt.

File: examples/rpc/starknet/starknet_addInvokeTransaction.hurl
Comment (glihm) [inline examples/rpc/starknet/starknet_addInvokeTransaction.hurl:11]: I need nonce set to 0 to be able to send the transaction as the first transaction.
Not sure if it's a problem from my setup, or something I am missing.

**Review by glihm — COMMENTED** [review]
Here the mapping in the client works well, and I am exploring the addition of receipts + how types converters must be done as @drspacemn did in his PR. It's a lot of mapping, investigating if there is an easier way to go here.

**Comment by ClementWalter** [comment]
I think that this is blocked until #452 is merged

**Comment by glihm** [comment]
> I think that this is blocked until #452 is merged

It is yes. :+1: 
#459 is the guy now, right?

**Comment by glihm** [comment]
Unlocked by merge of #459, I'll tackle this after work today. :+1:

**Comment by glihm** [comment]
@abdelhamidbakhta @tdelabro @EvolveArt 

On this PR, here is my current state:

1. I am waiting the `pendingTransactions` to be able to re-use the function they did to verify if the transaction is in the transaction pool.
2. On my tests, I have good behavior using hurl files. But in integration testing, I am doing exactly as `getTransactionByBlockIdAndIndex` to submit a transaction, but I can't retrieve it. Also I added a little code to get the tx hash back, but not sure it is the right way.

About the receipts, I have a situation.

1. The receipts are generated when a transaction is executed [here in the runtime](https://github.com/keep-starknet-strange/madara/blob/e664aeb879f957951f9331bc4e526ac0c8e547ce/crates/pallets/starknet/src/lib.rs#L509). After that, it is added to the `Pending` storage.
2. However, as we've seen, this storage is not populated until the block is init / execute txs / finalize. Our block interval time is 6s, this process occurs at like 5.8s. So it's super fast, and the probability we hit the RPC at this time is super small (at least for now).
3. So, during all this time, the transaction is in the transaction pool, where there is no receipt available.

I have talked with @drspacemn, here there is a confusion and I am not sure I'm getting it.

1. When the tx is in the transaction pool, it's status is supposed to be "Received".
2. When the tx is executed by the runtime (which for now it's a very short moment), and stored in `Pending` storage, status should be "Pending" (and a receipt is available).
3. When it's in a finalized block -> "accepted on L2".
4. (I'll skip "accepted on L1" for now)

I am not sure being in the right track as it's still a bit confused, please let me know your take on this.

But if I understood correctly substrate, we have here a transaction status that is missing. Because saying it's pending in the transaction pool, would be incorrect as no receipt is available. But we can stick to it right now, and the receipt will return "NoTxnHashFound".

**Comment by tdelabro** [comment]
@LucasLvy @abdelhamidbakhta 

> But if I understood correctly substrate, we have here a transaction status that is missing. Because saying it's pending in the transaction pool, would be incorrect as no receipt is available.

I don't understand, how could there be a receipt when the tx is still in the pool? Is it how Starknet works right now?

File: crates/client/rpc/src/lib.rs
Comment (tdelabro) [inline crates/client/rpc/src/lib.rs:701]: Using iterators and `find` is probably more idiomatic. Is there a reason why not to use it?

File: crates/pallets/starknet/src/lib.rs
Comment (tdelabro) [inline crates/pallets/starknet/src/lib.rs:879]: What is the problem here exactly?

Did you tried using `iter` rather than `into_iter` and only cloning in the end?

File: crates/client/rpc/src/lib.rs
Comment (tdelabro) [inline crates/client/rpc/src/lib.rs:763]: Can probably be factored with the previous one

File: crates/primitives/starknet/src/block/mod.rs
Comment (tdelabro) [inline crates/primitives/starknet/src/block/mod.rs:33]: I'm really not sure about this.
https://alpha-mainnet.starknet.io/feeder_gateway/get_block?blockNumber=latest

The receipts seem to be in a separate field of the `Block` struct. Not stored together with the transaction.

@LucasLvy any insight?

**Review by tdelabro — CHANGES_REQUESTED** [review]

File: crates/primitives/starknet/src/block/mod.rs
Comment (LucasLvy) [reply crates/primitives/starknet/src/block/mod.rs:33]: Yes they are different and are not in the same field

**Comment by LucasLvy** [comment]
> @LucasLvy @abdelhamidbakhta
> 
> > But if I understood correctly substrate, we have here a transaction status that is missing. Because saying it's pending in the transaction pool, would be incorrect as no receipt is available.
> 
> I don't understand, how could there be a receipt when the tx is still in the pool? Is it how Starknet works right now?

We introduced the pending state because blocks are really long on mainnet (could go up to 2 hours) so we needed a difference between tx in the mempool (received), tx processed but block not closed (pending) and tx included in a mined block (accepted on l2). As madara will have fast enough blocks we can ignore the pending state. The goal here is to return the receipt/information of mined txs. Once again let's completely ignore the pending state. Does that help ?

File: crates/client/db/src/mapping_db.rs
Comment (glihm) [reply crates/client/db/src/mapping_db.rs]: Good point you mentioned.
It was solved with the refacto with everything in the save commitment function like frontier.

File: crates/client/mapping-sync/src/sync_blocks.rs
Comment (glihm) [reply crates/client/mapping-sync/src/sync_blocks.rs]: So much cleaner yes, thanks for pointing this out.

File: crates/client/rpc/src/lib.rs
Comment (glihm) [reply crates/client/rpc/src/lib.rs]: I was limited by the fact that I didn't understand how to manage the possible error in the find. I have tried an implementation in the new commit with find.

File: crates/pallets/starknet/src/lib.rs
Comment (glihm) [reply crates/pallets/starknet/src/lib.rs]: I have changed here to support the two fields `transactions` and `transaction_receipts` of the block. I have made one for loop to ensure we only iterate one time on `pending`. Once again, I didn't manage to get this with `iter` and `map`. Let me know in the new commit if it's ok. :+1:

File: crates/client/rpc/src/lib.rs
Comment (glihm) [reply crates/client/rpc/src/lib.rs]: Refactored using find as you suggested.

**Comment by glihm** [comment]
> > @LucasLvy @abdelhamidbakhta
> > > But if I understood correctly substrate, we have here a transaction status that is missing. Because saying it's pending in the transaction pool, would be incorrect as no receipt is available.
> > 
> > 
> > I don't understand, how could there be a receipt when the tx is still in the pool? Is it how Starknet works right now?
> 
> We introduced the pending state because blocks are really long on mainnet (could go up to 2 hours) so we needed a difference between tx in the mempool (received), tx processed but block not closed (pending) and tx included in a mined block (accepted on l2). As madara will have fast enough blocks we can ignore the pending state. The goal here is to return the receipt/information of mined txs. Once again let's completely ignore the pending state. Does that help ?

Helps a lot, thank you @LucasLvy.

Is this summary accurate then?

1. Tx in the mempool:
    * `getTransactionByHash` -> return the tx from the pool. (but no field to mentioned that it's "received").
    * `getTransactionReceipt` -> `TxnHashNotFound`? But how to say that's it's received? Or should I submit a PR to add a type like [PendingInvokeTransactionReceipt](https://github.com/xJonathanLEI/starknet-rs/blob/a6cbfa3be71e9f42532c393632bdf18ccb01f88a/starknet-core/src/types/codegen.rs#L757) to have one like `ReceivedTransactionReceipt`?
    
2. Tx in the block but block not finished -> skip for now as you said.

3. Tx in mined block -> Transaction + Receipt ok.

My only concern is about having the user to know that the transaction is received. Which, for now, I don't know how to do so.

Thanks again for all your feedbacks and comment guys, helped a lot.

File: crates/primitives/starknet/src/block/mod.rs
Comment (glihm) [reply crates/primitives/starknet/src/block/mod.rs]: Last commit separates the two entities in the block. :+1:

**Comment by LucasLvy** [comment]
1. For txs in the mempool you return an error `TxNotFound` or something like that
2. Yes also return an error of that kind
3. Yes exactly

**Comment by glihm** [comment]
Waiting for #293 to check pending transactions.

File: crates/client/db/src/mapping_db.rs
Comment (LucasLvy) [inline crates/client/db/src/mapping_db.rs:106]:
```suggestion
    /// * `transaction_hash` - the transaction hash to search for. H256 is used here because it's a
```

File: crates/client/rpc/src/lib.rs
Comment (LucasLvy) [inline crates/client/rpc/src/lib.rs:759]: don't check in the mempool, this is the job of `starknet_pendingTransaction`

**Review by LucasLvy — CHANGES_REQUESTED** [review]

File: crates/client/db/src/mapping_db.rs
Comment (glihm) [reply crates/client/db/src/mapping_db.rs]: Good catch, thanks!

File: crates/client/rpc/src/lib.rs
Comment (glihm) [reply crates/client/rpc/src/lib.rs]: Makes sense.. sorry not having this conclusion before.
Just pushed. :+1:

**Review by LucasLvy — APPROVED** [review]
lgtm

**Pull request merged**
